[Bug] Exporting a modpack using CKAN 1.27.2 on Debian 4.19.118-2 locks up CKAN

Background
----------
* Operating System: Debian 4.19.118-2
* CKAN Version: 1.27.2
* KSP Version: 1.9.1.2788

* Mono version: 5.18.0.240

**Have you made any manual changes to your GameData folder (i.e., not via CKAN)?**
Some minor cfg file nudges, nothing earth shaking

Problem
-------
When attempting to export a modpack, the tab comes up and I can edit the name, identifier, and description, but not the version dropdowns,  The window where the list of mods is supposed to appear does not populate and no buttons do anything other than reflecting a click by changing color momentarily.  One cannot cancel out as it behaves as all other buttons.  Using the OS window widget to close CKAN does not exit, though it does change color so is reflecting the click.  I have to kill the mono process to get CKAN to end, then delete the lock file (hope that is what I should do and all I should do)


**Steps to reproduce**
* Go to...Export Modpack
* Click on...Cancel, or Export, or X-window close widget
* See error

**Expected behavior**
Populates modlist, exports, doesn't hang

**Screenshots (if applicable)**



**CKAN error code (if applicable):**

```
PLACE CODE HERE !
```

> I did find that list eventually, but how do I use that to start moving my install to 1.10 other than manually?

You can import that list again on another instance via `File > Install from .ckan`.

I also found the cause of the problem in my case. It's a mod installed from a locally crafted .ckan file, not listed in the repository. If I remove the mod the export tab works as expected and shows all mods (and you can cancel too)

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
---> CKAN.ModuleNotFoundKraken: Module not found: LocalMod 
  at CKAN.Registry.LatestAvailable (System.String module, CKAN.Versioning.KspVersionCriteria ksp_version, CKAN.RelationshipDescriptor relationship_descriptor) [0x0002d] in /git/CKAN/Core/Registry/Registry.cs:527 
  at CKAN.EditModpack+<>c__DisplayClass7_0.<AddGroup>b__1 (CKAN.RelationshipDescriptor r) [0x00000] in /git/CKAN/GUI/Controls/EditModpack.cs:117 
  at System.Linq.Enumerable+SelectIPartitionIterator`2[TSource,TResult].PreallocatingToArray (System.Int32 count) [0x0001e] in <cf901b2fef6a46be8867198917d43970>:0 
  at System.Linq.Enumerable+SelectIPartitionIterator`2[TSource,TResult].ToArray () [0x00023] in <cf901b2fef6a46be8867198917d43970>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <cf901b2fef6a46be8867198917d43970>:0 
  at CKAN.EditModpack.AddGroup (System.Collections.Generic.List`1[T] relationships, System.Windows.Forms.ListViewGroup group, CKAN.IRegistryQuerier registry) [0x0001e] in /git/CKAN/GUI/Controls/EditModpack.cs:115 
  at CKAN.EditModpack.LoadRelationships (CKAN.IRegistryQuerier registry) [0x0008a] in /git/CKAN/GUI/Controls/EditModpack.cs:96 
  at CKAN.EditModpack+<>c__DisplayClass1_0.<LoadModule>b__0 () [0x00219] in /git/CKAN/GUI/Controls/EditModpack.cs:61 
  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
  at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <873b82372c1942cead3b4ea98a2d7306>:0 
```

Now I'm wondering whether it is also a mod not listed in the index in your case, or maybe a mod we deindexed since you installed it?
Can you upload a screenshot of your installed mods (e.g. when selecting the installed filter)?

> > I did find that list eventually, but how do I use that to start moving my install to 1.10 other than manually?
> 
> You can import that list again on another instance via `File > Install from .ckan`.
> 
> I also found the cause of the problem in my case. It's a mod installed from a locally crafted .ckan file, not listed in the repository. If I remove the mod the export tab works as expected and shows all mods (and you can cancel too)
> 
> ```
> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
> ---> CKAN.ModuleNotFoundKraken: Module not found: LocalMod 
>   at CKAN.Registry.LatestAvailable (System.String module, CKAN.Versioning.KspVersionCriteria ksp_version, CKAN.RelationshipDescriptor relationship_descriptor) [0x0002d] in /git/CKAN/Core/Registry/Registry.cs:527 
>   at CKAN.EditModpack+<>c__DisplayClass7_0.<AddGroup>b__1 (CKAN.RelationshipDescriptor r) [0x00000] in /git/CKAN/GUI/Controls/EditModpack.cs:117 
>   at System.Linq.Enumerable+SelectIPartitionIterator`2[TSource,TResult].PreallocatingToArray (System.Int32 count) [0x0001e] in <cf901b2fef6a46be8867198917d43970>:0 
>   at System.Linq.Enumerable+SelectIPartitionIterator`2[TSource,TResult].ToArray () [0x00023] in <cf901b2fef6a46be8867198917d43970>:0 
>   at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <cf901b2fef6a46be8867198917d43970>:0 
>   at CKAN.EditModpack.AddGroup (System.Collections.Generic.List`1[T] relationships, System.Windows.Forms.ListViewGroup group, CKAN.IRegistryQuerier registry) [0x0001e] in /git/CKAN/GUI/Controls/EditModpack.cs:115 
>   at CKAN.EditModpack.LoadRelationships (CKAN.IRegistryQuerier registry) [0x0008a] in /git/CKAN/GUI/Controls/EditModpack.cs:96 
>   at CKAN.EditModpack+<>c__DisplayClass1_0.<LoadModule>b__0 () [0x00219] in /git/CKAN/GUI/Controls/EditModpack.cs:61 
>   at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
>   at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <873b82372c1942cead3b4ea98a2d7306>:0 
> ```
> 
> Now I'm wondering whether it is also a mod not listed in the index in your case, or maybe a mod we deindexed since you installed it?
> Can you upload a screenshot of your installed mods (e.g. when selecting the installed filter)?

Things are in flux here.  Not sure the modlist is the same now.  I'm working between 4 different test installs while trying to get a working one that my current game will have all it needs parts and modwise.  But now that you mention it, i do know that I manually installed an Infernal Robotics Next beta version that I'd hoped would fix something, I thought I'd removed it but there it is

--- a/Core/Registry/IRegistryQuerier.cs
+++ b/Core/Registry/IRegistryQuerier.cs
@@ -116,7 +116,7 @@ namespace CKAN
         /// This includes DLLs, which will have a version type of `DllVersion`.
         /// This includes Provides if set, which will have a version of `ProvidesVersion`.
         /// </summary>
-        Dictionary<string, ModuleVersion> Installed(bool include_provides = true);
+        Dictionary<string, ModuleVersion> Installed(bool withProvides = true, bool withDLLs = true);
 
         /// <summary>
         /// Returns the InstalledModule, or null if it is not installed.

--- a/Core/Registry/Registry.cs
+++ b/Core/Registry/Registry.cs
@@ -904,20 +904,23 @@ namespace CKAN
         /// <summary>
         /// <see cref = "IRegistryQuerier.Installed" />
         /// </summary>
-        public Dictionary<string, ModuleVersion> Installed(bool withProvides = true)
+        public Dictionary<string, ModuleVersion> Installed(bool withProvides = true, bool withDLLs = true)
         {
             var installed = new Dictionary<string, ModuleVersion>();
 
-            // Index our DLLs, as much as we dislike them.
-            foreach (var dllinfo in installed_dlls)
+            if (withDLLs)
             {
-                installed[dllinfo.Key] = new UnmanagedModuleVersion(null);
+                // Index our DLLs, as much as we dislike them.
+                foreach (var dllinfo in installed_dlls)
+                {
+                    installed[dllinfo.Key] = new UnmanagedModuleVersion(null);
+                }
             }
 
             // Index our provides list, so users can see virtual packages
             if (withProvides)
             {
-                foreach (var provided in Provided())
+                foreach (var provided in ProvidedByInstalled())
                 {
                     installed[provided.Key] = provided.Value;
                 }
@@ -948,14 +951,16 @@ namespace CKAN
         }
 
         /// <summary>
-        /// Returns a dictionary of provided (virtual) modules, and a
-        /// ProvidesVersion indicating what provides them.
+        /// Find modules provided by currently installed modules
         /// </summary>
-
-        // TODO: In the future it would be nice to cache this list, and mark it for rebuild
-        // if our installed modules change.
-        internal Dictionary<string, ProvidesModuleVersion> Provided()
+        /// <returns>
+        /// Dictionary of provided (virtual) modules and a
+        /// ProvidesVersion indicating what provides them
+        /// </returns>
+        internal Dictionary<string, ProvidesModuleVersion> ProvidedByInstalled()
         {
+            // TODO: In the future it would be nice to cache this list, and mark it for rebuild
+            // if our installed modules change.
             var installed = new Dictionary<string, ProvidesModuleVersion>();
 
             foreach (var modinfo in installed_modules)
@@ -1001,7 +1006,7 @@ namespace CKAN
             // withProvides is false.
             if (!with_provides) return null;
 
-            var provided = Provided();
+            var provided = ProvidedByInstalled();
 
             ProvidesModuleVersion version;
             return provided.TryGetValue(modIdentifier, out version) ? version : null;

--- a/Core/Registry/RegistryManager.cs
+++ b/Core/Registry/RegistryManager.cs
@@ -449,8 +449,19 @@ namespace CKAN
                 download_content_type = "application/zip",
             };
 
-            List<RelationshipDescriptor> mods = registry.Installed()
-                .Where(mod => !(mod.Value is ProvidesModuleVersion || mod.Value is UnmanagedModuleVersion))
+            List<RelationshipDescriptor> mods = registry.Installed(false, false)
+                .Where(kvp => {
+                    // Skip unavailable modules (custom .ckan files)
+                    try
+                    {
+                        var avail = registry.LatestAvailable(kvp.Key, null, null);
+                        return true;
+                    }
+                    catch
+                    {
+                        return false;
+                    }
+                })
                 .Select(kvp => (RelationshipDescriptor) new ModuleRelationshipDescriptor()
                     {
                         name    = kvp.Key,

--- a/GUI/Controls/EditModpack.cs
+++ b/GUI/Controls/EditModpack.cs
@@ -90,6 +90,20 @@ namespace CKAN
             }
 
             ignored.Clear();
+            // Find installed modules that aren't in the module's relationships
+            ignored.AddRange(registry.Installed(false, false)
+                .Where(kvp => {
+                    var ids = new string[] { kvp.Key };
+                    return !module.depends.Any(rel => rel.ContainsAny(ids))
+                        && !module.recommends.Any(rel => rel.ContainsAny(ids))
+                        && !module.suggests.Any(rel => rel.ContainsAny(ids));
+                })
+                .Select(kvp => (RelationshipDescriptor) new ModuleRelationshipDescriptor()
+                    {
+                        name    = kvp.Key,
+                        version = kvp.Value,
+                    })
+            );
             RelationshipsListView.Items.Clear();
             AddGroup(module.depends,    DependsGroup,         registry);
             AddGroup(module.recommends, RecommendationsGroup, registry);

--- a/GUI/Controls/EditModpack.cs
+++ b/GUI/Controls/EditModpack.cs
@@ -1,10 +1,8 @@
 using System;
-using System.IO;
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
 using System.Windows.Forms;
-using System.Text.RegularExpressions;
 using Autofac;
 using CKAN.Versioning;
 using CKAN.GameVersionProviders;
@@ -118,7 +116,9 @@ namespace CKAN
                         {
                             (r as ModuleRelationshipDescriptor)?.name,
                             (r as ModuleRelationshipDescriptor)?.version?.ToString(),
-                            registry.LatestAvailable((r as ModuleRelationshipDescriptor)?.name, null, null)?.@abstract
+                            registry.InstalledModules.First(
+                                im => im.identifier == (r as ModuleRelationshipDescriptor)?.name
+                            )?.Module.@abstract
                         })
                         {
                             Tag   = r,

--- a/GUI/Main/Main.cs
+++ b/GUI/Main/Main.cs
@@ -521,16 +521,6 @@ namespace CKAN
                         continue;
                     }
 
-                    // Don't add metapacakges to the registry
-                    if (!module.IsMetapackage)
-                    {
-                        // Remove this version of the module in the registry, if it exists.
-                        registry_manager.registry.RemoveAvailable(module);
-
-                        // Sneakily add our version in...
-                        registry_manager.registry.AddAvailable(module);
-                    }
-
                     if (module.IsDLC)
                     {
                         currentUser.RaiseError(Properties.Resources.MainCantInstallDLC, module);

